test/performance/odp_ml_perf.c

```diff
@@ -0,0 +1,1051 @@
+
+	if (odph_options(&helper_options)) {
+		ODPH_ERR("Failed to read ODP helper options.\n");
+		exit(EXIT_FAILURE);
+	}
+
+	/* List features not to be used */
+	odp_init_param_init(&init);
+	init.not_used.feat.cls = 1;
+	init.not_used.feat.compress = 1;
+	init.not_used.feat.crypto = 1;
+	init.not_used.feat.dma = 1;
+	init.not_used.feat.ipsec = 1;
+	init.not_used.feat.schedule = 1;
+	init.not_used.feat.timer = 1;
+	init.not_used.feat.tm = 1;
+	init.mem_model = helper_options.mem_model;
+
+	/* Init ODP before calling anything else */
+	if (odp_init_global(&inst, &init, NULL)) {
+		ODPH_ERR("Global init failed.\n");
+		exit(EXIT_FAILURE);
+	}
+
+	/* Init this thread */
+	if (odp_init_local(inst, ODP_THREAD_CONTROL)) {
+		ODPH_ERR("Local init failed.\n");
+		exit(EXIT_FAILURE);
+	}
+
+	odp_sys_info_print();
+
+	shm_glb = odp_shm_reserve("test_globals", sizeof(test_global_t), ODP_CACHE_LINE_SIZE, 0);
+
+	if (shm_glb != ODP_SHM_INVALID)
+		glb = (test_global_t *)odp_shm_addr(shm_glb);
+
+	if (!glb) {
+		ODPH_ERR("Failed to reserve shm\n");
+		ret = -1;
+		goto odp_term;
+	}
+
+	memset(glb, 0, sizeof(test_global_t));
+	parse_args(argc, argv);
+
+	if (read_file_to_shm(&glb->model_file_shm, &glb->model_file_data, &glb->model_file_size,
+			     glb->opt.model_name, "ml_perf_model")) {
+		ret = -1;
+		goto odp_term;
+	}
+
+	if (read_file_to_shm(&glb->inp_file_shm, &glb->inp_file_data, &glb->inp_file_size,
+			     glb->opt.input_name, "ml_perf_input")) {
+		ret = -1;
+		goto odp_term;
+	}
+
+	if (glb->opt.reference_name) {
+		if (read_file_to_shm(&glb->ref_file_shm, &glb->ref_file_data, &glb->ref_file_size,
+				     glb->opt.reference_name, "ml_perf_reference")) {
+			ret = -1;
+			goto odp_term;
+		}
+	}
+
+	if (odp_ml_capability(&glb->capa)) {
+		ODPH_ERR("odp_ml_capability() failed\n");
+		ret = -1;
+		goto odp_term;
+	}
+
+	if (glb->capa.max_model_size < glb->model_file_size) {
+		ODPH_ERR("Model size %" PRIu64 " exceeds maximum %" PRIu64 "\n",
+			 glb->model_file_size, glb->capa.max_model_size);
+		ret = -1;
+		goto odp_term;
+	}
+
+	if (glb->capa.min_input_align > 1) {
+		ODPH_ERR("Minimum input alignment %d not supported\n", glb->capa.min_input_align);
+		ret = -1;
+		goto odp_term;
+	}
+
+	if (glb->capa.min_output_align > 1) {
+		ODPH_ERR("Minimum output alignment %d not supported\n", glb->capa.min_output_align);
+		ret = -1;
+		goto odp_term;
+	}
+
+	odp_ml_config_t ml_config;
+	odp_ml_model_param_t model_param;
+
+	odp_ml_config_init(&ml_config);
+	ml_config.max_model_size = glb->capa.max_model_size;
+	ml_config.load_mode_mask = ODP_ML_COMPL_MODE_SYNC;
+	ml_config.run_mode_mask = ODP_ML_COMPL_MODE_SYNC;
+
+	if (odp_ml_config(&ml_config)) {
+		ODPH_ERR("odp_ml_config() failed\n");
+		ret = -1;
+		goto odp_term;
+	}
+
+	odp_ml_model_param_init(&model_param);
+	model_param.model = glb->model_file_data;
+	model_param.size = glb->model_file_size;
+
+	glb->mdl = odp_ml_model_create(glb->opt.model_name, &model_param);
```

Comment:
The model created in the main seems to remain created and loaded during the test runs. This might have some result skewing effects? Tester should work in all modes also in the case of `odp_ml_capability_t::max_models == 1`.

Reply:
Seems also that, if the core count that is used for testing is more than supported `max_models` ("Create-destroy" mode), tester just fails without any information what is the maximum allowed count. There should some capability checks at the beginning to check these and inform user. No point starting the test runs if there is zero chance of success.

Reply:
Added a couple if checks for max_models*.